src/preprocessing/numerical.rs

```diff
@@ -0,0 +1,406 @@
+    /// should be kept. Otherwise, replace it by zero.
+    fn adjust_column_mean(&self, mean: T) -> T {
+        if self.parameters.with_mean {
+            mean
+        } else {
+            T::zero()
+        }
+    }
+    /// When the standard-deviation should be adjusted, the column
+    /// standard-deviation should be kept. Otherwise, replace it by zero.
+    fn adjust_column_std(&self, std: T) -> T {
+        if self.parameters.with_std {
+            ensure_std_valid(std)
+        } else {
+            T::one()
+        }
+    }
+}
+
+/// Make sure the standard deviation is valid. If it is
+/// negative or zero, it should replaced by the smallest
+/// positive value the type can have. That way we can savely
+/// divide the columns with the resulting scalar.
+fn ensure_std_valid<T: RealNumber>(value: T) -> T {
+    value.max(T::min_positive_value())
+}
+
+/// During `fit` the `StandardScaler` computes the column means and standard deviation.
+impl<T: RealNumber, M: Matrix<T>> UnsupervisedEstimator<M, StandardScalerParameters>
+    for StandardScaler<T>
+{
+    fn fit(x: &M, parameters: StandardScalerParameters) -> Result<Self, Failed> {
+        Ok(Self {
+            means: x.column_mean(),
+            stds: x.std(0),
+            parameters,
+        })
+    }
+}
+
+/// During `transform` the `StandardScaler` applies the summary statistics
+/// computed during `fit` to set the mean of each column to zero and the
+/// standard deviation to one.
+impl<T: RealNumber, M: Matrix<T>> Transformer<M> for StandardScaler<T> {
+    fn transform(&self, x: &M) -> Result<M, Failed> {
+        let (_, n_cols) = x.shape();
+        if n_cols != self.means.len() {
+            return Err(Failed::because(
+                FailedError::TransformFailed,
+                &format!(
+                    "Expected {} columns, but got {} columns instead.",
+                    self.means.len(),
+                    n_cols,
+                ),
+            ));
+        }
+
+        Ok(build_matrix_from_columns(
+            self.means
+                .iter()
+                .zip(self.stds.iter())
+                .enumerate()
+                .map(|(column_index, (column_mean, column_std))| {
+                    take_column_from_matrix(x, column_index)
+                        .sub_scalar(self.adjust_column_mean(*column_mean))
+                        .div_scalar(self.adjust_column_std(*column_std))
+                })
+                .collect(),
+        )
+        .unwrap())
+    }
+}
+
+/// Extract a single column from a matrix and return it as a matrix.
+fn take_column_from_matrix<T, M>(matrix: &M, column_index: usize) -> M
+where
+    T: RealNumber,
+    M: Matrix<T>,
+{
+    matrix.take(&[column_index], 1)
+}
+
+/// From a collection of matrices, that contain columns, construct
+/// a matrix by stacking the columns horizontally.
+fn build_matrix_from_columns<T, M>(columns: Vec<M>) -> Option<M>
```

Comment:
Why this is a `Vec<M>` instead of a `Vec<Vec<T>>`>

Reply:
The `build_matrix_from_columns` function was build by me as a helper / utility function in this module. During `fit` of the `StandardScaler` I first take individual columns from the matrix that contains the raw data then apply the necessary scaling. In short I need `Vec<M>` instead of `Vec<Vec<T>>` because `take_column_from_matrix` returns a `Matrix`. Also this is helpful because I can use the neat methods `sub_scalar`  as well as `div_scaler` to apply the scaling.

Reply:
ok but for it to be general it is better to use `Vec<Vec<T>>` as you won't need build da matrix from a matrix but a matrix from a 2-D vector. So if you plan to make `build_matrix_from_columns` generic it would be better for it to accept Vecs

Reply:
Ok that make sense. Then I would prefer not to put `build_matrix_from_columns` into `Matrix`  because I would need to refactor it and it would not be optimal for this module anymore.